dev/bots/prepare_package.dart

```diff
@@ -61,131 +66,314 @@ class ArchiveCreator {
     }
     args.addAll(<String>[
       output.absolute.path,
       path.basename(source.absolute.path),
     ]);
 
-    _runProcess(zip, args,
+    return _runProcess(zip, args,
         workingDirectory: new Directory(path.dirname(source.absolute.path)));
   }
 
-  void createTarArchive(File output, Directory source) {
+  Future<String> createTarArchive(File output, Directory source) {
     final List<String> args = <String>[
       'cJf',
       output.absolute.path,
       path.basename(source.absolute.path),
     ];
-    _runProcess(tar, args,
+    return _runProcess(tar, args,
         workingDirectory: new Directory(path.dirname(source.absolute.path)));
   }
+
+  /// Finds the largest "plain" version number (just #.#.#, no alpha or symbols)
+  /// that is not greater than the current version by sorting the version numbers
+  /// (including the current one) in descending order according to release order
+  /// and returning the one before the current one.
+  String _getLargestPreviousVersion(String currentVersion, List<String> tags) {
+    assert(tags != null && tags.isNotEmpty);
+    final RegExp normalReleaseRe = new RegExp(r'^([0-9]+)\.([0-9]+)\.([0-9]+)(-dev)?$');
+    tags.add(currentVersion);
+    tags.sort((String a, String b) {
+      final Match matchA = normalReleaseRe.firstMatch(a);
+      final Match matchB = normalReleaseRe.firstMatch(b);
+      // Make sure non-release (and non-normal release tags) get sorted at the bottom
+      // of the list.
+      if (matchA == null) {
+        return 1;
+      }
+      if (matchB == null) {
+        return -1;
+      }
+      final int majorDiff = int.parse(matchB.group(1)) - int.parse(matchA.group(1));
+      if (majorDiff != 0) {
+        return majorDiff;
+      }
+      final int minorDiff = int.parse(matchB.group(2)) - int.parse(matchA.group(2));
+      if (minorDiff != 0) {
+        return minorDiff;
+      }
+      final int revDiff = int.parse(matchB.group(3)) - int.parse(matchA.group(3));
+      if (revDiff != 0) {
+        return revDiff;
+      }
+      final int devA = matchA.group(4) == null ? 0 : 1;
+      final int devB = matchB.group(4) == null ? 0 : 1;
+      return devB - devA;
+    });
+    return tags[tags.indexOf(currentVersion) + 1];
+  }
+
+  /// Gets the output name of the archive.  The output name is a combination
+  /// of "flutter_" with the OS, the current version in the VERSION file, and the
+  /// number of commits since that version changed appended to it.
+  ///
+  /// So, for example, when the VERSION file contains '0.0.21-dev', if the
+  /// current revision has 85 commits since the last alpha roll, and we're on Linux,
+  /// the string returned from this function is:
+  ///
+  /// flutter_linux_0.0.21-dev.85.tar.xz
+  ///
+  /// When the version number in VERSION doesn't include '-dev', i.e. when we're
+  /// doing an alpha roll, then it is tagged with alpha instead, giving:
+  ///
+  /// flutter_linux_0.0.22.alpha.tar.xz
```

Comment:
While we are in the process of rolling alpha (and VERSION doesn't have a dev suffix) people can still commit PRs. In that case, we would end up with multiple flutter_linux_0.0.22.alpha.tar.xz all named the same but with different content. Is that something to worry about? Which one is the "real" alpha release?

Reply:
When I change this to use the tags, then that won't be a problem, since there can be only one commit that is tagged.  I suppose the issue then is that the alpha isn't tagged yet when we roll alpha (it's tagged retroactively), so maybe I'll just abandon the alpha labeling here, and we can rename the archive as part of the alpha roll process instead.

Reply:
Let's chat on Monday to figure out what the best naming is and how this should work with the new release process. It may be that we want to just create these with the git hash as the name, and then later retroactively rename them when we're doing a release/tagging a commit.